Add ESC50 and zero-shot classification

RahulSChand opened the pull request:
For #2069 

<!-- If you are submitting a dataset or a model for the model registry please use the corresponding checklists below otherwise feel free to remove them. -->

<!-- add additional description, question etc. related to the new dataset -->


### Code Quality
<!-- Please do not delete this -->
- [ ✅ ] **Code Formatted**: Format the code using `make lint` to maintain consistent style.

### Documentation
<!-- Please do not delete this -->
- [ ] **Updated Documentation**: Add or update documentation to reflect the changes introduced in this PR.

### Testing
<!-- Please do not delete this -->
- [ ] **New Tests Added**: Write tests to cover new functionality. Validate with `make test-with-coverage`.
- [ ] **Tests Passed**: Run tests locally using `make test` or `make test-with-coverage` to ensure no existing functionality is broken.


### Adding datasets checklist
<!-- see also https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md -->

**Reason for dataset addition**: ... <!-- Add reason for adding dataset here. E.g. it covers task/language/domain previously not covered -->

- [ ✅] I have run the following models on the task (adding the results to the pr). These can be run using the `mteb -m {model_name} -t {task_name}` command.
  - [ ] `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`
  - [ ] `intfloat/multilingual-e5-small`
- [ ✅] I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).
- [ ] If the dataset is too big (e.g. >2048 examples), considering using `self.stratified_subsampling() under dataset_transform()`
- [✅ ] I have filled out the metadata object in the dataset file (find documentation on it [here](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md#2-creating-the-metadata-object)).
- [ ] Run tests locally to make sure nothing is broken using `make test`. 
- [ ✅] Run the formatter to format the code using `make lint`. 


### Adding a model checklist
<!-- 
When adding a model to the model registry
see also https://github.com/embeddings-benchmark/mteb/blob/main/docs/reproducible_workflow.md
-->

 - [ ✅] I have filled out the ModelMeta object to the extent possible
 - [✅ ] I have ensured that my model can be loaded using
   - [ ✅] `mteb.get_model(model_name, revision)` and
   - [ ]✅ `mteb.get_model_meta(model_name, revision)`
 - [✅ ] I have tested the implementation works on a representative set of tasks.

KennethEnevoldsen commented:
Hei @RahulSChand it would be great if you would fill out the checklist - that makes sure that everything works. Also do add a description of the changes (it makes it easier to review)

We should also enable CI tests when merging into 

I especially see some errors that the lining would resolve (use of `Path` instead of `os.join`).

RahulSChand commented:
Sure, I will fill out the checklist and add a description.

RahulSChand commented:
Hi @KennethEnevoldsen I updated the PR with end to end testing of the ESC50 dataset & laion/clap-htsat-unfused model for zero-shot classification. I get 81.7% zero-shot accuracy. I used the following snippet to test the end to end flow. 

Things remaining here are 1) Adding new test cases 2) And ensuring `make test` passes 3) Any comments from ur side

Snippet to test zero-shot audio classification

```
from __future__ import annotations

import mteb

print("mteb imported.")

model_name = "laion/clap-htsat-unfused"
model = mteb.get_model(model_name=model_name)
print("model loaded..")
tasks = mteb.get_tasks(tasks=["ESC50_Zeroshot"])
print("task loaded..")

print(tasks)

evaluation = mteb.MTEB(tasks=tasks)

results = evaluation.run(model)
print("eval complete.")

print(results[0].scores)
```

![image](https://github.com/user-attachments/assets/982c87d2-e308-4d26-a706-19a19c877b19)

Samoed commented on mteb/tasks/Audio/AudioMultilabelClassification/eng/ESC50.py:
I think this file should be updated to be as multilabel classification

Samoed commented on mteb/abstasks/Audio/AbsTaskAudioMultilabelClassification.py:
Can you update init, train/eval spilts similarly as was done in `v2`? https://github.com/embeddings-benchmark/mteb/blob/v2.0.0/mteb/abstasks/AbsTaskMultilabelClassification.py

Samoed commented on mteb/abstasks/Audio/AbsTaskZeroshotClassificaiton.py:
```diff
@@ -0,0 +1,64 @@
+        )
+        metrics = evaluator(model, encode_kwargs=encode_kwargs)
+
+        scores = {"accuracy": metrics["accuracy"]}
```
Can you add f1 scores?

Samoed commented on mteb/abstasks/Audio/AbsTaskZeroshotClassificaiton.py:
```diff
@@ -0,0 +1,64 @@
+
+class AbsTaskZeroshotClassification(AbsTask):
+    """Abstract class for ZeroshotClassification tasks
+    The similarity between an image and candidate text prompts, such as this is a .
```
Can you continue docstring?
```suggestion
    The similarity between an image and candidate text prompts, such as this is a ....
```

RahulSChand commented:
@Samoed Hey, I finally got around to fixing this PR. I was waiting for this other PR #2082  to merge since many file changes were similar. I have updated the PR with

1) removed all unnecessary files
2) Removed esc50 from multi-label task, its only in zero-shot for now
3) Added F1 score 
4) Made sure make test passes
5) Ran the code end to end using the below script

Currently on the fused clap model is there, when this PR gets merged, I will create another small PR with just the unfused model added as well. 

```python
import mteb
model_name = "laion/clap-htsat-fused"
model = mteb.get_model(model_name=model_name)

tasks = mteb.get_tasks(tasks=["ESC50_Zeroshot"])

evaluation = mteb.MTEB(tasks=tasks)
results = evaluation.run(model)

print(results[0].scores)
```

![Screenshot from 2025-03-05 12-22-17](https://github.com/user-attachments/assets/2819f4e4-c5c6-409f-815d-741e211ba727)

Samoed commented on mteb/evaluation/evaluators/Audio/ZeroshotClassificationEvaluator.py:
```diff
@@ -0,0 +1,127 @@
+        """Evaluate zero-shot classification performance."""
+        logger.info("Getting text embeddings for candidate labels...")
+
+        print("Available methods for model:")
+        for method in dir(model):
+            if not method.startswith('_'):  # Skip private/internal methods
+                print(f"- {method}")
+
+        print(type(model))
+        
+        print(model,type(model))
+
+        # assert False
```
```suggestion
```

RahulSChand replied:
Fixed

Samoed commented on mteb/models/clap_models.py:
```diff
@@ -0,0 +1,210 @@
+        **kwargs: Any,
+    ):
+        self.model_name = model_name
+        self.device = device
+        self.model = ClapModel.from_pretrained(model_name).to(self.device)
+        self.processor = ClapProcessor.from_pretrained(model_name)
+        self.sampling_rate = self.processor.feature_extractor.sampling_rate
+
+        self.pipeline = pipeline(
+            task="zero-shot-audio-classification", model=model_name, device=device
+        )
```
Does this load 2 versions of model? One `self.model` and second `pipeline`. Can you pass model directly to pipeline? I think this can reduce memory usage

Samoed replied:
```suggestion
        self.device = device
        self.model = ClapModel.from_pretrained(model_name).to(self.device)
        self.processor = ClapProcessor.from_pretrained(model_name)
        self.sampling_rate = self.processor.feature_extractor.sampling_rate

        self.pipeline = pipeline(
            task="zero-shot-audio-classification", model=model_name, device=device, **kwargs
        )
```

RahulSChand replied:
Fixed the **kwargs. Will check the pipeline thing

RahulSChand replied:
Yeah the model was double-loading, fixed it + tested if it still works

Samoed commented on mteb/tasks/Audio/__init__.py:
```diff
@@ -1,4 +1,8 @@
+<<<<<<< HEAD
+from .AudioZeroshotClassification import *
+=======
+>>>>>>> upstream/maeb
```
```suggestion
from .AudioZeroshotClassification import *
```

Samoed commented on tests/test_benchmark/mock_tasks.py:
```diff
@@ -581,7 +584,78 @@ def load_data(self, **kwargs):
         )
         self.data_loaded = True
 
+<<<<<<< HEAD
```
Here also some chnages from merge conflict

Samoed commented on mteb/evaluation/evaluators/Audio/ZeroshotClassificationEvaluator.py:
```diff
@@ -0,0 +1,127 @@
+        predictions = similarity.argmax(dim=1).cpu().numpy()
+
+        # Calculate metrics
+        scores = {
```
I think precision and recall can also be added

RahulSChand replied:
Added

Samoed reviewed (commented):
Great job! Here are a few small suggestions

Samoed commented on mteb/models/clap_models.py:
```diff
@@ -0,0 +1,214 @@
+    revision="main",
+    release_date="2023-05-22",
+    modalities=["audio", "text"],
+    n_parameters=None,  # Fill in actual number
+    memory_usage_mb=None,  # Fill in actual number
```
Missed that

RahulSChand replied:
Fixed
